### PR TITLE
Add Kingdom-internal Panel Match data-layer gRPC service definitions.

### PR DESCRIPTION
**Pull request by efoxepstein**
For now, this adds a minimal set of methods required to implement most
of the functionality of the public facing APIs.

As more functionality is required, these will need to be expanded.
Building them incrementally has the advantage of not needing to
implement methods that are later determined to be unnecessary.

See world-federation-of-advertisers/cross-media-measurement#3.

---
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10)

### CODE REVIEW

File: src/main/proto/wfa/measurement/internal/kingdom/exchanges_service.proto [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  // If there is already an Exchange for the same date for the given
+  // `RecurringExchange`, this will do nothing and return that instead.
+  rpc CreateExchange(CreateExchangeRequest) returns (Exchange);
+}
+
+// Request for `/Exchanges.CreateExchange`.
+message CreateExchangeRequest {
+  // The `Exchange` to create.
+  Exchange exchange = 1;
+}
```
Comment (kungfucraig) [inline src/main/proto/wfa/measurement/internal/kingdom/exchanges_service.proto:23]: newline

**Review by kungfucraig — APPROVED** [review]

**Review by efoxepstein — COMMENTED** [review]
> Dismissed @kungfucraig from a discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10)* status: 0 of 7 files reviewed, all discussions resolved (waiting on @SanjayVas)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchanges_service.proto, line 23 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10#-MXmDZ7hvdMbkDtcWoVf-r1-23:-MXmEJLc9LEjy61meYTX:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/16bfe25f165f27953da349b9f124f62371c0a5c2/src/main/proto/wfa/measurement/internal/kingdom/exchanges_service.proto#L23)):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

newline
</blockquote></details>

Done.

**Review by SanjayVas — COMMENTED** [review]
> Reviewed 7 of 7 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10)* status: all files reviewed, 3 unresolved discussions (waiting on @efoxepstein)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange.proto, line 38 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10#-MXmDn0W4qeNtN4sljl5:-MXmDn0W4qeNtN4sljl6:bhbl2o4) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/16bfe25f165f27953da349b9f124f62371c0a5c2/src/main/proto/wfa/measurement/internal/kingdom/exchange.proto#L38)):*
> ```
> proto
> ```
nit: Use "message" as the short term in .proto files for "protocol buffer message" rather than "proto".
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 64 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10#-MXmEDQw0Hz86wNrqYlW:-MXmEDQw0Hz86wNrqYlX:b-peo1lg) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/16bfe25f165f27953da349b9f124f62371c0a5c2/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L64)):*
> ```
> proto
> ```
nit: same here
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempts_service.proto, line 47 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10#-MXmHc-GE5rJwtgIL7k_:-MXmHc-GE5rJwtgIL7ka:bzajgh) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/16bfe25f165f27953da349b9f124f62371c0a5c2/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempts_service.proto#L47)):*
> ```
> fixed64
> ```
int32 is used for step indices elsewhere.

As a side note: this isn't universal, but I generally see the Latin pluralization of indices used in mathematical contexts (e.g. index of a vector) and the modified English pluralization for other contexts (e.g. the index at the end of a book, which is similar to the concept of a DB index). I don't care as long as there's some rule we're following, e.g. always using the modified English pluralization regardless of context.

**Review by efoxepstein — COMMENTED** [review]
> Dismissed @SanjayVas from a discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @efoxepstein)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange.proto, line 38 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10#-MXmDn0W4qeNtN4sljl5:-MXmKCMBCrd1MO4INsgk:bb6un1j) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/16bfe25f165f27953da349b9f124f62371c0a5c2/src/main/proto/wfa/measurement/internal/kingdom/exchange.proto#L38)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

> ```
> proto
> ```
nit: Use "message" as the short term in .proto files for "protocol buffer message" rather than "proto".
</blockquote></details>

Done
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 64 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10#-MXmEDQw0Hz86wNrqYlW:-MXmK9e8B5m_QtgzPVQ2:bb6un1j) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/16bfe25f165f27953da349b9f124f62371c0a5c2/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L64)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

> ```
> proto
> ```
nit: same here
</blockquote></details>

Done
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempts_service.proto, line 47 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10#-MXmHc-GE5rJwtgIL7k_:-MXmJrjT3fNTyEcMoOYb:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/16bfe25f165f27953da349b9f124f62371c0a5c2/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempts_service.proto#L47)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

> ```
> fixed64
> ```
int32 is used for step indices elsewhere.

As a side note: this isn't universal, but I generally see the Latin pluralization of indices used in mathematical contexts (e.g. index of a vector) and the modified English pluralization for other contexts (e.g. the index at the end of a book, which is similar to the concept of a DB index). I don't care as long as there's some rule we're following, e.g. always using the modified English pluralization regardless of context.
</blockquote></details>

Done.

**Review by SanjayVas — APPROVED** [review]
> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @efoxepstein)

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto [hunk 2]
```diff
@@ -38,8 +38,30 @@ message ExchangeStepAttempt {
...
   // subsequent attempt should increase the last `attempt_number` by 1.
+  // Output-only.
   int32 attempt_number = 5;
 
   // Additional metadata about the ExchangeStepAttempt.
   ExchangeStepAttemptDetails details = 6;
+
+  enum State {
+    STATE_UNSPECIFIED = 0;
+
```
Comment (stevenwarejones) [inline src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto:49]: Airflow has the following states, I believe:
```
Dagrun: SUCCESS, RUNNING, FAILED;
Task: SUCCESS, RUNNING, FAILED, UPSTREAM_FAILED, SKIPPED, UP_FOR_RETRY, UP_FOR_RESCHEDULE, QUEUED, NONE, SCHEDULED.
```
I don't want to overthink this but I could see `up_for_retry` and `queued` being helpful status states

File: src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto [hunk 3]
```diff
@@ -38,8 +38,30 @@ message ExchangeStepAttempt {
 
   // Which attempt this is. The first attempt should have index 1 and each
   // subsequent attempt should increase the last `attempt_number` by 1.
+  // Output-only.
   int32 attempt_number = 5;
 
   // Additional metadata about the ExchangeStepAttempt.
   ExchangeStepAttemptDetails details = 6;
+
+  enum State {
```
Comment (stevenwarejones) [inline src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto:47]: `ExchangeStepAttemptDetails` also can contain a state. What is the difference?

**Review by efoxepstein — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10)* status: all files reviewed, 2 unresolved discussions (waiting on @kunyuchen and @stevenwarejones)

___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 47 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10#-MXmDZ7hvdMbkDtcWoVc-r3-47:-MYCUbcX4nnZqVPS2VKD:by7y994) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/038f96cedf43b0e9b6a8e22ddc893f8a41bd2194/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L47)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

`ExchangeStepAttemptDetails` also can contain a state. What is the difference?
</blockquote></details>

This PR moves the state from ExchangeStepAttemptDetails to ExchangeStepAttempt.

The general philosophy is to shove everything into Details except that which will be indexed or queried on. This minimizes the number of database columns to just those necessary. Before this CL it wasn't clear that the state was important to model directly in the database schema. However, with this CL, we add a filter on the StreamExchangeStepAttempts gRPC method and thus we'll likely want to index on the state. I can add the database schema change in this PR or in a future one. Not super important to me.
___
*[src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto, line 49 at r3](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/10#-MXmDZ7hvdMbkDtcWoVc-r3-49:-MYCWDxr8ulyfdC7bwNj:b-oc6qdk) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/038f96cedf43b0e9b6a8e22ddc893f8a41bd2194/src/main/proto/wfa/measurement/internal/kingdom/exchange_step_attempt.proto#L49)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

Airflow has the following states, I believe:
```
Dagrun: SUCCESS, RUNNING, FAILED;
Task: SUCCESS, RUNNING, FAILED, UPSTREAM_FAILED, SKIPPED, UP_FOR_RETRY, UP_FOR_RESCHEDULE, QUEUED, NONE, SCHEDULED.
```
I don't want to overthink this but I could see `up_for_retry` and `queued` being helpful status states
</blockquote></details>

In the spirit of YAGNI (https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it), let's keep the minimal set of states until we discover uses for more.

**Review by stevenwarejones — APPROVED** [review]

**Pull request merged**